airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeDatabase.java

```diff
@@ -4,58 +4,191 @@
-        .build();
-
-    return Databases.createBasicDataSource(username, password, jdbcUrl.toString(), DRIVER_CLASS_NAME, properties);
+    dataSource.setDriverClassName(DRIVER_CLASS_NAME);
+    dataSource.setJdbcUrl(jdbcUrl.toString());
+    dataSource.setDataSourceProperties(properties);
+    return dataSource;
+  }
+
+  private static String getAccessTokenUsingRefreshToken(final String hostName,
+                                                        final String clientId,
+                                                        final String clientSecret,
+                                                        final String refreshCode)
+      throws IOException {
+    final var refreshTokenUri = String.format(REFRESH_TOKEN_URL, hostName);
+    final Map<String, String> requestBody = new HashMap<>();
+    requestBody.put("grant_type", "refresh_token");
+    requestBody.put("refresh_token", refreshCode);
+
+    try {
+      final BodyPublisher bodyPublisher = BodyPublishers.ofString(requestBody.keySet().stream()
+          .map(key -> key + "=" + URLEncoder.encode(requestBody.get(key), StandardCharsets.UTF_8))
+          .collect(joining("&")));
+
+      final byte[] authorization = Base64.getEncoder()
+          .encode((clientId + ":" + clientSecret).getBytes(StandardCharsets.UTF_8));
+
+      final HttpRequest request = HttpRequest.newBuilder()
+          .POST(bodyPublisher)
+          .uri(URI.create(refreshTokenUri))
+          .header("Content-Type", "application/x-www-form-urlencoded")
+          .header("Accept", "application/json")
+          .header("Authorization", "Basic " + new String(authorization, StandardCharsets.UTF_8))
+          .build();
+
+      final HttpResponse<String> response = httpClient.send(request,
+          HttpResponse.BodyHandlers.ofString());
+
+      final JsonNode jsonResponse = Jsons.deserialize(response.body());
+      if (jsonResponse.has("access_token")) {
+        return jsonResponse.get("access_token").asText();
+      } else {
+        throw new RuntimeException(
+            "Failed to obtain accessToken using refresh token. " + jsonResponse);
+      }
+    } catch (final InterruptedException e) {
+      throw new IOException("Failed to refreshToken", e);
+    }
   }
 
   public static JdbcDatabase getDatabase(final JsonNode config) {
     final DataSource dataSource = createDataSource(config);
     return new DefaultJdbcDatabase(dataSource);
   }
 
+  private static Runnable getRefreshTokenTask(final HikariDataSource dataSource){
+    return () -> {
+      LOGGER.info("Refresh token process started");
+      var props = dataSource.getDataSourceProperties();
+      try {
+        var token = getAccessTokenUsingRefreshToken(props.getProperty("host"),
+            props.getProperty("client_id"), props.getProperty("client_secret"),
+            props.getProperty("refresh_token"));
+        props.setProperty("token", token);
+
+        LOGGER.info("New refresh token has been obtained");
```

Comment:
the new refresh token is never persisted back with the config though? does it need to be propagated all the way back to the airbyte DB?

Reply:
updated, thanks